### DISCUSSION TITLE
浅析 Bigtable 和 LevelDB 的实现 · /bigtable-leveldb

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#253] [Announcements]
https://draveness.me/bigtable-leveldb

**Comment by zurex** [comment]
想问下作者图使用什么画的， 很好看，排版也很不错

**Comment by draveness** [comment]
> @zurex
> 想问下作者图使用什么画的， 很好看，排版也很不错

Sketch

**Comment by zurex** [comment]
> @Draveness
> > @zurex
> > 想问下作者图使用什么画的， 很好看，排版也很不错
> 
> Sketch

谢谢

**Comment by tmckobe** [comment]
好赞！楼主看完leveldb代码大概用了多久？

**Comment by mugbya** [comment]
> 这与现在大多的数据库的实现完全相同，通过顺序写向日志追加记录，然后再向数据库随机写，因为随机写的耗时远远大于追加内容，如果直接进行随机写，可能由于发生设备故障造成数据丢失。

随机写的耗时远远大于追加内容? 这个我可能懂得太少，有些不能理解，博主方便详细解释下么，还有我贴出这句话，感觉主语确实，读起来很矛盾呐

**Comment by zurex** [comment]
> @mugbya
> > 这与现在大多的数据库的实现完全相同，通过顺序写向日志追加记录，然后再向数据库随机写，因为随机写的耗时远远大于追加内容，如果直接进行随机写，可能由于发生设备故障造成数据丢失。
> 
> 随机写的耗时远远大于追加内容? 这个我可能懂得太少，有些不能理解，博主方便详细解释下么，还有我贴出这句话，感觉主语确实，读起来很矛盾呐
> 
> 

这个术语叫做WAL(Write-Ahead Logging), 随机写相对于追加写有着额外的磁盘寻道开销。
其实原文这里有些错误，设备故障导致的数据丢失不论是随机写还是追加写都无法避免，单纯的磁盘故障一般使用raid来解决。

**Comment by draveness** [comment]
> @zurex
> > @mugbya
> > > 这与现在大多的数据库的实现完全相同，通过顺序写向日志追加记录，然后再向数据库随机写，因为随机写的耗时远远大于追加内容，如果直接进行随机写，可能由于发生设备故障造成数据丢失。
> > 
> > 随机写的耗时远远大于追加内容? 这个我可能懂得太少，有些不能理解，博主方便详细解释下么，还有我贴出这句话，感觉主语确实，读起来很矛盾呐
> > 
> > 
> 
> 这个术语叫做WAL(Write-Ahead Logging), 随机写相对于追加写有着额外的磁盘寻道开销。
> 其实原文这里有些错误，设备故障导致的数据丢失不论是随机写还是追加写都无法避免，单纯的磁盘故障一般使用raid来解决。

非常感谢提出的问题，这里想要表达的意思其实是，因为随机写消耗的时间大于顺序写，所以写入时发生磁盘故障最终导致数据丢失的**概率**会更高，而不是说使用顺序写就不会发生磁盘故障，我在文中加入了『由于随机写执行时间较长』，希望尽量不会误导后面的读者。

**Comment by mugbya** [comment]
> @zurex
> > @mugbya
> > > 这与现在大多的数据库的实现完全相同，通过顺序写向日志追加记录，然后再向数据库随机写，因为随机写的耗时远远大于追加内容，如果直接进行随机写，可能由于发生设备故障造成数据丢失。
> > 
> > 随机写的耗时远远大于追加内容? 这个我可能懂得太少，有些不能理解，博主方便详细解释下么，还有我贴出这句话，感觉主语确实，读起来很矛盾呐
> > 
> > 
> 
> 这个术语叫做WAL(Write-Ahead Logging), 随机写相对于追加写有着额外的磁盘寻道开销。
> 其实原文这里有些错误，设备故障导致的数据丢失不论是随机写还是追加写都无法避免，单纯的磁盘故障一般使用raid来解决。

感谢这么热心，你的订阅包括回复也会通知到你 ？

**Comment by mugbya** [comment]
> @draveness
> > @zurex
> > > @mugbya
> > > > 这与现在大多的数据库的实现完全相同，通过顺序写向日志追加记录，然后再向数据库随机写，因为随机写的耗时远远大于追加内容，如果直接进行随机写，可能由于发生设备故障造成数据丢失。
> > > 
> > > 随机写的耗时远远大于追加内容? 这个我可能懂得太少，有些不能理解，博主方便详细解释下么，还有我贴出这句话，感觉主语确实，读起来很矛盾呐
> > > 
> > > 
> > 
> > 这个术语叫做WAL(Write-Ahead Logging), 随机写相对于追加写有着额外的磁盘寻道开销。
> > 其实原文这里有些错误，设备故障导致的数据丢失不论是随机写还是追加写都无法避免，单纯的磁盘故障一般使用raid来解决。
> 
> 非常感谢提出的问题，这里想要表达的意思其实是，因为随机写消耗的时间大于顺序写，所以写入时发生磁盘故障最终导致数据丢失的**概率**会更高，而不是说使用顺序写就不会发生磁盘故障，我在文中加入了『由于随机写执行时间较长』，希望尽量不会误导后面的读者。

OK，感谢博主，现在理解了。再次感谢，话说博主头像看起来像德莱文啊

**Comment by draveness** [comment]
> @mugbya
>
> 感谢这么热心，你的订阅包括回复也会通知到你 ？

会给我发邮件的

**Comment by zurex** [comment]
> @mugbya
> > @zurex
> > > @mugbya
> > > > 这与现在大多的数据库的实现完全相同，通过顺序写向日志追加记录，然后再向数据库随机写，因为随机写的耗时远远大于追加内容，如果直接进行随机写，可能由于发生设备故障造成数据丢失。
> > > 
> > > 随机写的耗时远远大于追加内容? 这个我可能懂得太少，有些不能理解，博主方便详细解释下么，还有我贴出这句话，感觉主语确实，读起来很矛盾呐
> > > 
> > > 
> > 
> > 这个术语叫做WAL(Write-Ahead Logging), 随机写相对于追加写有着额外的磁盘寻道开销。
> > 其实原文这里有些错误，设备故障导致的数据丢失不论是随机写还是追加写都无法避免，单纯的磁盘故障一般使用raid来解决。
> 
> 感谢这么热心，你的订阅包括回复也会通知到你 ？

是的。这个评论背后是github的issues, 会自动发送提醒到我的邮箱

**Comment by SimonCqk** [comment]
`唯一与缓存不同可能就是，在数据『命中』后，它并不会把数据移动到更近的地方，而是会把数据移到更远的地方来减少下一次的访问时间`
这里不是很懂，指的是read在SSTable中命中key，同时触发了compaction后，key对应的SSTable到了level更高的地方吗？

**Comment by GardianT** [comment]
bloom filter相关逻辑没有介绍？也算是比较重要的特性吧？

**Comment by GardianT** [comment]
还有skip list相关的，加上这两个介绍的话，真的就是比较全面的leveldb的介绍了

**Comment by 00arthur00** [comment]
文章很赞，在看第三遍。
并且将其中一部分标记未删除的信息彻底清除
有个字好像写错了：
并且将其中一部分标记`为`删除的信息彻底清除

**Comment by draveness** [comment]
> 文章很赞，在看第三遍。
> 并且将其中一部分标记未删除的信息彻底清除
> 有个字好像写错了：
> 并且将其中一部分标记`为`删除的信息彻底清除

修复了

**Comment by holmes1412** [comment]
你好～文章写得非常清晰！获益良多！
想问下“多层级的 SSTable”部分，第二张图中，4个sstable是否应该是“level0”而不是“level1”呢？

**Comment by draveness** [comment]
> 你好～文章写得非常清晰！获益良多！
> 想问下“多层级的 SSTable”部分，第二张图中，4个sstable是否应该是“level0”而不是“level1”呢？

感觉差不太多，这里就不改了

**Comment by itliusir** [comment]
hello 试了下Sketch 感觉没那个味 大佬可以把平时用的素材库/模板分享一份么

**Comment by hhhizzz** [comment]
写的很好 获益匪浅 但有个地方还是想说一下
> 如果这次 Get 方法在某个层级的 SSTable 中找到了对应的键，那么就算做上一层级中包含该键的 SSTable 的一次查找

这里应该是算是上一层级的缺失？因为是在上一层级没找到才进入下一层级找‘allowed_seeks’ 应该是指缺失的次数 如果在这个sstable里找数据 每次都找不到而在更高层找到了说明他应该和更高层的数据合并
这么说的话最后说把命中数据挪到更远不太对吧？ 命中的数据并没有被挪远 挪远的是没有被命中的 从而使得这个数据下次更容易被命中了

**Comment by fanyingjie11** [comment]
"为了减轻主服务器的负载，所有的客户端仅仅通过 Master 获取 tablet 服务器的位置信息"，这句话是有问题的，客户端不是通过master获取tablet服务器的位置信息的，原文有这么一句话：“Because Bigtable clients also do not rely on the master for tablet location information, most clients never communicate with the master”，望悉知！
